datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -1799,4 +1803,34 @@ mod test {
         assert!(result.len() == 1);
         Ok(())
     }
+
+    #[test]
+    fn test_short_circuits() -> Result<()> {
+        let table_scan = test_table_scan()?;
+
+        let extracted_short_circuit = col("a").eq(lit(0)).or(col("b").eq(lit(0)));
+        let not_extracted_short_circuit_leg = (col("a") + col("b")).eq(lit(0));
+        let plan = LogicalPlanBuilder::from(table_scan.clone())
+            .project(vec![
+                extracted_short_circuit.clone().alias("c1"),
+                extracted_short_circuit.alias("c2"),
+                col("c")
+                    .gt(lit(0))
+                    .or(not_extracted_short_circuit_leg.clone())
+                    .alias("c3"),
+                col("c")
+                    .gt(lit(1))
+                    .or(not_extracted_short_circuit_leg)
+                    .alias("c4"),
+            ])?
+            .build()?;
+
+        let expected = "Projection: __common_expr_1 AS c1, __common_expr_1 AS c2, test.c > Int32(0) OR test.a + test.b = Int32(0) AS c3, test.c > Int32(1) OR test.a + test.b = Int32(0) AS c4\
+        \n  Projection: test.a = Int32(0) OR test.b = Int32(0) AS __common_expr_1, test.a, test.b, test.c\
+        \n    TableScan: test";
+
+        assert_optimized_plan_eq(expected, plan, None);
+
+        Ok(())
+    }
```

Comment:
Can we add a test case like the one below to check if (a or b) can be extracted as a common subexpr?
```
select ((a or b) or d) as f1, ((a or b) or c) as f2 from t;
```

Reply:
This PR doesn't extract surely evaluated children of short circuiting expressions so I kept that `TODO` yet ( https://github.com/apache/datafusion/pull/11197/files#diff-351499880963d6a383c92e156e75019cd9ce33107724a9635853d7d4cd1898d0R955).
I will address that in a separate follow-up PR...

Reply:
It would be nice to add the test (as a negative test perhaps)

Reply:
I've adjusted the test case to test both legs of an OR expression: https://github.com/apache/datafusion/pull/11197/commits/8b75d82fbd837be723a43cb4edd6cf88369e4d18. In this PR none of them are extraceted, but the after that follow-up PR the srurely evaluated first leg (called `not_extracted_short_circuit_leg_1` now) will be extracted.